src/arti/types/pandas.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import annotations
+
+from typing import Any, cast
+
+import numpy as np
+import pandas as pd
+
+from arti.types import List, String, Struct, Type, TypeAdapter, TypeSystem
+from arti.types.numpy import numpy_type_system
+
+# TODO: How should (multi)indexes be handled; perhaps as a "hint"?
+
+pandas_type_system = TypeSystem(key="pandas", extends=(numpy_type_system,))
+
+
+@pandas_type_system.register_adapter
+class SeriesAdapter(TypeAdapter):
+    artigraph = List
+    system = pd.Series
+
+    @classmethod
+    def matches_artigraph(cls, type_: Type, *, hints: dict[str, Any]) -> bool:
+        return (
+            isinstance(type_, cls.artigraph)
+            # List(element=Struct(...)) are handled by the DataFrameAdapter.
+            and not isinstance(type_.element, Struct)
+        )
+
+    @classmethod
+    def to_artigraph(cls, type_: Any, *, hints: dict[str, Any], type_system: TypeSystem) -> Type:
+        dtype = type_.dtype
+        if dtype == np.dtype("O"):
+            # TODO: Should we handle empty series by defaulting to "String", but issuing
+            # a warning?
+            example_value = type_.iloc[0]
+            if isinstance(example_value, str):
+                return List(element=String())
+            # TODO: Handle dicts, lists, etc.
+            raise NotImplementedError(
+                f"Non-string {dtype} is not supported yet, got values of: {example_value}"
+            )
+        return List(element=type_system.to_artigraph(dtype, hints=hints))
+
+    @classmethod
+    def matches_system(cls, type_: Any, *, hints: dict[str, Any]) -> bool:
+        return isinstance(type_, cls.system)
+
+    @classmethod
+    def to_system(cls, type_: Type, *, hints: dict[str, Any], type_system: TypeSystem) -> Any:
+        assert isinstance(type_, cls.artigraph)
+        dtype = type_system.to_system(type_.element, hints=hints)
+        return pd.Series([dtype()], dtype=dtype)
+
+
+@pandas_type_system.register_adapter
+class DataFrameAdapter(TypeAdapter):
```

Comment:
How does this handle a series of objects?

Reply:
That part is handled in the `SeriesAdapter`; `dtype("O")` strings will be handled alright, but other types currently raise an error:
https://github.com/artigraph/artigraph/blob/fef1e9905114a7b1b4d45f6cde47d4968afe0c9b/src/arti/types/pandas.py#L32-L41

I think for complex cell values, we may want to try to convert into a python type annotation (eg: `list[str]` or `list[dict[str, str]]`) and pass off to the `python_type_system`. I'm not quite sure whether that "value -> type annotation" logic should live here or in the `python_type_system` though (probably leaning towards the latter with some sort of "infer_type" logic?).

Reply:
FYI: I added a (failing) test and https://github.com/artigraph/artigraph/issues/258 for this.